src/stream.rs

```diff
@@ -0,0 +1,216 @@
+    }
+}
+
+impl<S: TryStream, B: Backoff, Sl: Sleeper> Stream for StreamBackoff<S, B, Sl>
+where
+    Sl::Sleep: Future,
+{
+    type Item = Result<S::Ok, S::Error>;
+
+    fn poll_next(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        match this.state.as_mut().project() {
+            StateProj::BackingOff { mut backoff_sleep } => match backoff_sleep.as_mut().poll(cx) {
+                Poll::Ready(()) => {
+                    // tracing::debug!(deadline = ?backoff_sleep.deadline(), "Backoff complete, waking up");
+                    this.state.set(State::Awake)
+                }
+                Poll::Pending => {
+                    // let deadline = backoff_sleep.deadline();
+                    // tracing::trace!(
+                    //     ?deadline,
+                    //     remaining_duration = ?deadline.saturating_duration_since(Instant::now()),
+                    //     "Still waiting for backoff sleep to complete"
+                    // );
+                    return Poll::Pending;
+                }
+            },
+            StateProj::GivenUp => {
+                // tracing::debug!("Backoff has given up, stream is closed");
+                return Poll::Ready(None);
+            }
+            StateProj::Awake => {}
+        }
+
+        let next_item = this.stream.try_poll_next(cx);
+        match &next_item {
+            Poll::Ready(Some(Err(_))) => {
+                if let Some(backoff_duration) = this.backoff.next_backoff() {
+                    let backoff_sleep = this.sleeper.sleep(backoff_duration);
+                    // tracing::debug!(
+                    //     deadline = ?backoff_sleep.deadline(),
+                    //     duration = ?backoff_duration,
+                    //     "Error received, backing off"
+                    // );
+                    this.state.set(State::BackingOff { backoff_sleep });
+                } else {
+                    // tracing::debug!("Error received, giving up");
+                    this.state.set(State::GivenUp);
+                }
+            }
+            Poll::Ready(_) => {
+                // tracing::trace!("Non-error received, resetting backoff");
+                this.backoff.reset();
+            }
+            Poll::Pending => {}
+        }
+        next_item
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::StreamBackoff;
+    use crate::{backoff::Backoff, future::TokioSleeper};
+    use futures_channel::mpsc;
+    use futures_util::{pin_mut, poll, stream, StreamExt};
+    use std::{task::Poll, time::Duration};
+    use tokio_1 as tokio;
+
+    #[tokio::test]
+    async fn stream_should_back_off() {
+        tokio::time::pause();
+        let tick = Duration::from_secs(1);
+        let rx = stream::iter([Ok(0), Ok(1), Err(2), Ok(3), Ok(4)]);
+        let rx = StreamBackoff::new(rx, crate::backoff::Constant::new(tick), TokioSleeper);
```

Comment:
Need to use Tokio specifically for these marble tests, since async-std doesn't support time travel (and, regardless, `Sleeper` doesn't and probably shouldn't support that).